src/main/java/org/truffleruby/extra/ffi/PointerNodes.java

```diff
@@ -130,46 +142,63 @@ public DynamicObject add(DynamicObject a, long b) {
     @Primitive(name = "pointer_read_string", lowerFixnum = 1)
     public static abstract class PointerReadStringPrimitiveNode extends PrimitiveArrayArgumentsNode {
 
+        public static PointerReadStringPrimitiveNode create() {
+            return PointerNodesFactory.PointerReadStringPrimitiveNodeFactory.create(null);
+        }
+
         @Specialization
         public DynamicObject readString(DynamicObject pointer, int length) {
             final byte[] bytes = new byte[length];
-            Layouts.POINTER.getPointer(pointer).get(0, bytes, 0, length);
+            readPointer(Layouts.POINTER.getPointer(pointer), bytes, length);
             return createString(RopeBuilder.createRopeBuilder(bytes));
         }
 
+        @TruffleBoundary
+        private void readPointer(Pointer pointer, byte[] out, int length) {
+            pointer.get(0, out, 0, length);
+        }
     }
 
     @Primitive(name = "pointer_read_string_to_null")
     public static abstract class PointerReadStringToNullPrimitiveNode extends PrimitiveArrayArgumentsNode {
 
+        @Child private PointerReadStringPrimitiveNode readStringNode;
+
         @Specialization(guards = "isNullPointer(pointer)")
         public DynamicObject readNullPointer(DynamicObject pointer) {
             return createString(RopeConstants.EMPTY_ASCII_8BIT_ROPE);
         }
 
-        @TruffleBoundary
         @Specialization(guards = "!isNullPointer(pointer)")
         public DynamicObject readStringToNull(DynamicObject pointer) {
             if (TruffleOptions.AOT) {
-                final jnr.ffi.Pointer ptr = Layouts.POINTER.getPointer(pointer);
-                final int nullOffset = ptr.indexOf(0, (byte) 0);
-                final byte[] bytes = new byte[nullOffset];
+                if (readStringNode == null) {
+                    CompilerDirectives.transferToInterpreterAndInvalidate();
+                    readStringNode = insert(PointerReadStringPrimitiveNode.create());
+                }
 
-                ptr.get(0, bytes, 0, bytes.length);
+                final jnr.ffi.Pointer ptr = Layouts.POINTER.getPointer(pointer);
+                final int nullOffset = findNullOffset(ptr);
 
-                return StringOperations.createString(getContext(), RopeBuilder.createRopeBuilder(bytes));
+                return readStringNode.readString(pointer, nullOffset);
```

Comment:
It would be simpler to just make `readPointer` above a static method and probably name it readString or readBytes.
If you prefer to keep the node, then make a generic `executeX` method, never call specializations directly from another node as it breaks encapsulation.